src/transaction/atomic_replication_helper.hpp

```diff
@@ -169,36 +166,85 @@ namespace cublog
-	      int fix_page (THREAD_ENTRY *thread_p);
-	      void unfix_page (THREAD_ENTRY *thread_p);
-	      PAGE_PTR get_page_ptr ();
-	      void set_page_ptr (const PAGE_PTR &ptr);
-	      LOG_LSA get_lsa () const;
+	      void apply_log_by_type (THREAD_ENTRY *thread_p, log_rv_redo_context &redo_context,
+				      LOG_RECTYPE rectype) const;
 
 	      const VPID m_vpid;
 	    private:
 	      const log_lsa m_record_lsa;
 	      const LOG_RCVINDEX m_record_index;
-	      PAGE_PTR m_page_ptr;
+	      // ownership of page pointer is with the bookkeeper in the owning class; this is just a
+	      // reference to allow applying the redo function when needed
+	      PAGE_PTR const m_page_ptr;
+	  };
+
+	  using page_ptr_watcher_uptr_type = std::unique_ptr<PGBUF_WATCHER>;
+
+	  struct page_ptr_info
+	  {
+	    page_ptr_info () = default;
+
+	    page_ptr_info (const page_ptr_info &) = delete;
+	    page_ptr_info (page_ptr_info &&) = default;
+
+	    page_ptr_info &operator= (const page_ptr_info &) = delete;
+	    page_ptr_info &operator= (page_ptr_info &&) = delete;
 
-	      std::unique_ptr<PGBUF_WATCHER> m_watcher_p;
+	    ~page_ptr_info ();
+
+	    VPID m_vpid = VPID_INITIALIZER;
+	    LOG_RCVINDEX m_rcv_index = RV_NOT_DEFINED;
+	    PAGE_PTR m_page_p = nullptr;
+	    page_ptr_watcher_uptr_type m_watcher_p;
+	    int m_ref_count = -1;
+	  };
+
+	  /*
+	   * Implements a RAII-like reference counted functionality to bokkeep page pointers for
+	   * a sequence of [possibly] nested atomic replication sub-sequences.
+	   * A page can be needed by multiple levels of a nested atomic replication sequence which
+	   * perfom changes on the page. Once the page is unfixed in a sequence at a certain
+	   * level, it can be:
+	   *  - either still kept fixed if a parent [sub]sequence did the fixing and still
+	   *    needs the page
+	   *  - or unfixed if there is no parent [sub]sequence which needs the page anymore (aka:
+	   *    the [sub]sequence which just requested the fix is the outer-most one that needed
+	   *    the page in the current overall sequence of possibly nested [sub]sequences
+	   */
+	  struct page_ptr_bookkeeping
+	  {
+	    public:
+	      page_ptr_bookkeeping () = default;
+	      ~page_ptr_bookkeeping ();
+
+	      page_ptr_bookkeeping (const page_ptr_bookkeeping &) = delete;
+	      page_ptr_bookkeeping (page_ptr_bookkeeping &&) = delete;
+
+	      page_ptr_bookkeeping &operator= (const page_ptr_bookkeeping &) = delete;
+	      page_ptr_bookkeeping &operator= (page_ptr_bookkeeping &&) = delete;
+
+	    public: // methods
+	      int fix_page (THREAD_ENTRY *thread_p, VPID vpid, LOG_RCVINDEX rcv_index, PAGE_PTR &page_ptr_out);
+	      int unfix_page (THREAD_ENTRY *thread_p, VPID vpid);
+
+	    private: // types
+	      using page_ptr_info_map_type = std::map<VPID, page_ptr_info>;
+
+	    private: // variables
+	      page_ptr_info_map_type m_;
```

Comment:
It would be better to change it to a more meaningful name. Such as `m_page_ptr_info_map`

Reply:
Done